Add scoop functionality to EcoBin classes

graelo opened the pull request:
Hi, many thanks for this Addon, I love it. I was just missing the scoop for EcoBins.

This PR extends scoop support to both `EcoBin` and `CustomEcoBin` classes, bringing feature parity with standard storage bins.

## Changes

- EcoBin: Added scoop properties and shape generation logic
- CustomEcoBin: Added scoop properties with proper custom shape integration (scoop is trimmed by `inside_wall_negative`

Both classes now expose `Scoop` (boolean) and `ScoopRadius` (length) properties in the UI. Default scoop value set to False for backward compatibility

## Implementation

The implementation reuses the existing `make_scoop()` function and follows established patterns used by other features like label shelves. For `CustomEcoBin`, the scoop is properly cut by the inside wall negative to handle custom shapes correctly.

## Testing

- Code passes linting (ruff) and type checking (mypy)
- I could not identify any breaking changes to existing functionality
- Maintains backward compatibility with existing EcoBin objects

This addresses the feature gap where scoops were available in standard bins but missing from eco bins.

Here's an EcoBin with a scoop (radius: 12mm)

<img width="951" height="672" alt="image" src="https://github.com/user-attachments/assets/7d823c0a-180e-4552-bbe1-6eecc0f8ff0b" />

The underside is unaffected:
<img width="763" height="643" alt="image" src="https://github.com/user-attachments/assets/eb0f4223-f392-4a05-8a23-a079cf0ba056" />

It also works for 1x1 EcoBins
<img width="466" height="443" alt="image" src="https://github.com/user-attachments/assets/f30ae6fc-afe8-485b-9585-a472879ecc9b" />

and custom EcoBins
<img width="678" height="749" alt="image" src="https://github.com/user-attachments/assets/049c1752-6dd0-4765-8d57-64d7a07bdb8a" />

graelo committed:
feat: add scoop to ecobins

greg19 commented on freecad/gridfinity_workbench/features.py:
```diff
@@ -311,7 +313,21 @@ def generate_gridfinity_shape(self, obj: fc.DocumentObject) -> Part.Shape:
+
+        # Now add scoop, but only where eco compartments exist (reversed logic)
+        if obj.Scoop:
+            # Store original UsableHeight and adjust for eco bin's deeper compartment
+            original_usable_height = obj.UsableHeight
+            obj.UsableHeight = obj.TotalHeight - obj.BaseWallThickness
+            scoop = feat.make_scoop(obj)
+            obj.UsableHeight = original_usable_height  # Restore original value
```
```suggestion
            scoop = feat.make_scoop(obj, usable_height=obj.TotalHeight - obj.BaseWallThickness)
```
Can you modify the `make_scoop` to take the `UsableHeight` as an argument instead?

greg19 replied:
You can do
```py
def make_scoop(obj, *, usable_height=None):
    if usable_height is None:
        usable_height = obj.UsableHeight
```
If you don't want to add this argument in all the places this function is called

graelo replied:
Hi, thanks, this makes more sense than my little dance above.

It still works great

CustomEcoBin height 3 with full radius (21mm)

<img width="681" height="536" alt="image" src="https://github.com/user-attachments/assets/253456fb-fd2a-4f6e-ac13-025d9280c4f7" />

With constrained radius of 12 mm

<img width="696" height="564" alt="image" src="https://github.com/user-attachments/assets/b6668ab7-1464-49b5-84a5-815c04fa5515" />

greg19 reviewed (commented):
Nice, cool to have it, thanks for contributing :slightly_smiling_face:

graelo committed:
fix: ruff I001 import sorting and float(usable_height) unit mismatch

Add [tool.ruff.lint.isort] known-third-party config so FreeCAD, FreeCADGui,
and Part are classified consistently on macOS and Linux. Revert import
reordering to alphabetical within a single block.

Remove float() wrappers from usable_height in make_scoop() which stripped
FreeCAD Quantity units, causing "Unit mismatch in plus operation" at runtime.
Use targeted type: ignore[arg-type] comments for the freecad-stubs false
positives instead.

greg19 commented on freecad/gridfinity_workbench/feature_construction.py:
```diff
@@ -192,8 +195,20 @@ def scoop_properties(obj: fc.DocumentObject, *, scoop_default: bool) -> None:
 
-def make_scoop(obj: fc.DocumentObject) -> Part.Shape:
-    """Create scoop."""
+def make_scoop(obj: fc.DocumentObject, *, usable_height: None | int = None) -> Part.Shape:
```
```suggestion
def make_scoop(obj: fc.DocumentObject, *, usable_height: None | fc.Units.Quantity = None) -> Part.Shape:
```

graelo commented:
I'll rebase soon.

Stu142 committed:
version bump

Stu142 commented:
Sorry for the extended delay to merge it seems to work great. 

> I found that I had been having issues with Ruff import sorting locally due to the fact that macOS filesystem is case insensitive. On macOS (case-insensitive filesystem), ruff classifies FreeCAD as first-party because it matches the project's freecad/ directory. This causes it to expect a blank line between import Part (third-party) and import FreeCAD (first-party), and to sort Part before FreeCAD. On Linux CI (case-sensitive), both are third-party, so ruff expects them in the same block, alphabetically sorted — and flags the macOS ordering as I001.

Interesting I am on windows but I think others were linux so there may be something similar going on I'll have to look into it.

graelo commented:
Thanks!
